ambry-cloud/src/main/java/com/github/ambry/cloud/VcrReplicaThread.java

```diff
@@ -76,6 +81,44 @@ public VcrReplicaThread(String threadName, FindTokenHelper findTokenHelper, Clus
     this.azureCloudConfig = new AzureCloudConfig(properties);
     this.azureTableNameReplicaTokens = this.azureCloudConfig.azureTableNameReplicaTokens;
     this.azureMetrics = new AzureMetrics(clusterMap.getMetricRegistry());
+    this.numReplIter = 0;
+  }
+
+  class ReplicaComparator implements Comparator<RemoteReplicaInfo> {
+    @Override
+    public int compare(RemoteReplicaInfo r1, RemoteReplicaInfo r2) {
+      String d1 = r1.getReplicaId().getDataNodeId().getHostname();
+      String d2 = r2.getReplicaId().getDataNodeId().getHostname();
+      return d1.compareTo(d2);
+    }
+  }
+
+  /**
+   * A custom filter that inheritors can override depending upon use case to pick which replicas to replicate from
+   * @param replicas A map of replicas {host -> {replicas}}
+   */
+  @Override
+  public Map<DataNodeId, List<RemoteReplicaInfo>> customFilter(Map<DataNodeId, List<RemoteReplicaInfo>> replicas) {
+    HashMap<Long, List<RemoteReplicaInfo>> partitions = new HashMap<>();
+    // Group replicas by partition
+    replicas.values().forEach(rlist -> rlist.forEach(replica -> {
+      long pid = replica.getReplicaId().getPartitionId().getId();
+      List alist = partitions.getOrDefault(pid, new ArrayList<>());
```

Comment:
The code here is trying to group replicas by partition, so partition -> {r1, r2, r3}.
`alist` is not created each time we enter the loop, but only once when partition is not in the map.

When partition p is not in the map:
1. create new alist
2. add r1 to alist
3. insert p -> {r1}

When partition p is already in the map:
1. get existing alist
2. append r2 to alist
3. put nothing since p already exists